File: healthcheck/go/src/compute_health.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"strconv"
	"time"
)

func main() {
	checksFile := os.Args[1]
	checksInterval, err := strconv.Atoi(os.Getenv("INTERVAL"))
	check(err)

	targetFolder := os.Getenv("SUMMARY_DIR")

	checksJSON, err := ioutil.ReadFile(checksFile)
	check(err)

	var checks []HealthCheck
	err = json.Unmarshal(checksJSON, &checks)
	check(err)

	for {
		fmt.Println("Computing health")

		summary := make(map[string]*HealthSummary)
		allSummary := new(HealthSummary)
		allSummary.Status = 0
		allStartTime := time.Now()
		for _, check := range checks {
			fmt.Printf(" - %+v?\n", check.Description)

			checkSummary := new(HealthSummary)
			startTime := time.Now()

			// TODO: capture stdout --> https://stackoverflow.com/a/40770011/539599
			parameters := expandEnv(check.Command[1:])
			cmd := exec.Command(check.Command[0], parameters...)
			err := cmd.Run() // TODO: Use timout
			// --> https://medium.com/@vCabbage/go-timeout-commands-with-os-exec-commandcontext-ba0c861ed738

			if err != nil {
				fmt.Printf("   -> ERROR: %+v\n", err)
				checkSummary.Status = exitStatus(err)
				checkSummary.Output = "??" // TODO
				allSummary.Status = 1
			} else {
				fmt.Println("   -> OKAY")
				checkSummary.Status = 0
				checkSummary.Output = "??" // TODO
			}

			checkSummary.Duration = Duration(time.Now().Sub(startTime))
			summary[check.Description] = checkSummary
		}

		allSummary.Duration = Duration(time.Now().Sub(allStartTime))
		summary["all"] = allSummary

		summaryFile, err := json.MarshalIndent(summary, "", "    ")
		check(err)

		err = ioutil.WriteFile(targetFolder+"/all.json", summaryFile, os.ModeExclusive)
		check(err)
		err = os.Chmod(targetFolder+"/all.json", 0644)
		check(err)

		time.Sleep(time.Duration(checksInterval) * time.Second)
	}
}

File: healthcheck/ruby/src/compute_health.rb
require "bundler"
Bundler.require(:default)
require_relative 'shared.rb'

checks_interval = ENV['INTERVAL'].to_i
target_folder = ENV['SUMMARY_DIR']

checks_file = ARGV[0]
checks_json = File.read(checks_file)
checks = JSON.parse(checks_json, symbolize_names: true)

while true
    puts "Computing health"

    summary = { all: { status: 0 } }
    all_start_time = Time.now
    checks.each do |check|
        puts " - #{check[:description]}?"

        check_summary = {}
        start_time = Time.now

        # TODO: compute check
        sleep 1

        check_summary[:duration] = pretty_duration(Time.now, start_time)
        summary[check[:description]] = check_summary
    end
    summary[:all][:duration] = pretty_duration(Time.now, all_start_time)
    summary[:all][:finished] = Time.now

    File.open("#{target_folder}/all.json", 'w') { |f| f.write(JSON.pretty_generate(summary)) }
    FileUtils.chmod(0644, "#{target_folder}/all.json")

    sleep checks_interval
end
